File: TOKI/Book/01-PerkenalanPemrogramanKompetitif/A.LampuDanBotol/2.SolusiLanjut/Main.java
import java.util.*;

public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        long n = sc.nextInt();
        sc.close();

        long num = n;
        int divisorCount = 1;
        for(long i = 2; i*i <= num; i++) {
            int factorCount = 0;
            while(num % i == 0) {
                factorCount++;
                num /= i;
            }

            divisorCount *= (1 + factorCount);
        }
        if (num > 1) { // Sisa faktor
            divisorCount *= 2;
        }

        if(divisorCount % 2 == 0) {
            System.out.println("lampu mati");
        }else {
            System.out.println("lampu menyala");
        }
    }
}

File: TOKI/Book/01-PerkenalanPemrogramanKompetitif/A.LampuDanBotol/2.SolusiLanjut/main.cpp
#include <iostream>
using namespace std;

int main(){
    long long N;
    cin >> N;
    long long num = N;
    
    int divisorCount = 1;
    for (long long i = 2; i * i <= num; i++) {
        int factorCount = 0;
        while (num % i == 0) {
            factorCount++;
            num /= i;
        }
        divisorCount *= (1 + factorCount);
    }

    if (num > 1) { // Sisa faktor
        divisorCount *= 2;
    }

    if (divisorCount % 2 == 0) {
        cout << "lampu mati" << endl;
    } else {
        cout << "lampu menyala" << endl;
    }
}